examples/ramberg_osgood.py

```diff
@@ -0,0 +1,666 @@
+    sorted_facets = np.argsort(entity_indices)
+    mesh_tags = df.mesh.meshtags(domain, edim, entity_indices[sorted_facets], entity_markers[sorted_facets])
+    return mesh_tags, marked
+
+
+def build_view(cells, V):
+    mesh = V.mesh
+    submesh, cell_map, _, _ = df.mesh.create_submesh(mesh, mesh.topology.dim, cells)
+    fe = V.ufl_element()
+    V_sub = df.fem.FunctionSpace(submesh, fe)
+
+    submesh = V_sub.mesh
+    view_parent = []
+    view_child = []
+
+    num_sub_cells = submesh.topology.index_map(submesh.topology.dim).size_local
+    for cell in range(num_sub_cells):
+        view_child.append(V_sub.dofmap.cell_dofs(cell))
+        view_parent.append(V.dofmap.cell_dofs(cell_map[cell]))
+    if view_child:
+        return (
+            np.hstack(view_parent),
+            np.hstack(view_child),
+            V_sub,
+        )
+    else:
+        # it may be that a process does not own any of the cells in the submesh
+        return (
+            np.array([], dtype=np.int32),
+            np.array([], dtype=np.int32),
+            V_sub,
+        )
+
+
+class IncrSmallStrainProblem(df.fem.petsc.NonlinearProblem):
+    def __init__(
+        self,
+        # laws = [(law_1, cells_1), (law_2, cells_2), ...]
+        laws: list[tuple[Any, np.ndarray]],
+        u: df.fem.Function,
+        q_degree: int = 2,
+    ):
+        mesh = u.function_space.mesh
+        map_c = mesh.topology.index_map(mesh.topology.dim)
+        num_cells = map_c.size_local + map_c.num_ghosts
+        self.cells = np.arange(0, num_cells, dtype=np.int32)
+
+        self.gdim = mesh.ufl_cell().geometric_dimension()
+        QVe = ufl.VectorElement(
+            "Quadrature", mesh.ufl_cell(), q_degree, quad_scheme="default", dim=6
+        )
+        QTe = ufl.TensorElement(
+            "Quadrature",
+            mesh.ufl_cell(),
+            q_degree,
+            quad_scheme="default",
+            shape=(6, 6),
+        )
+        QV = df.fem.FunctionSpace(mesh, QVe)
+        QT = df.fem.FunctionSpace(mesh, QTe)
+
+        self.laws = []
+        self.QV_views = []
+        self.QT_views = []
+        self._strain = []
+        self._stress = []
+        self._tangent = []
+
+        with df.common.Timer("submeshes-and-data-structures"):
+            for law, cells in laws:
+                self.laws.append((law, cells))
+
+                # ### submesh and subspace for strain, stress
+                QV_parent, QV_child, QV_sub = build_view(cells, QV)
+                self.QV_views.append((QV_parent, QV_child, QV_sub))
+                self._strain.append(df.fem.Function(QV_sub))
+                self._stress.append(df.fem.Function(QV_sub))
+
+                # ### submesh and subspace for tanget
+                QT_parent, QT_child, QT_sub = build_view(cells, QT)
+                self.QT_views.append((QT_parent, QT_child, QT_sub))
+                self._tangent.append(df.fem.Function(QT_sub))
+
+        self.stress = df.fem.Function(QV)
```

Comment:
If we want this problem definition to work with incremental laws, we need the stress two times. One time for the updated state and one for the initial state. Same for history variables